libdnf/utils/fs/file.hpp

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_UTILS_FS_FILE_HPP
+#define LIBDNF_UTILS_FS_FILE_HPP
+
+#include <filesystem>
+
+
+namespace libdnf::utils::fs {
+
+/// A wrapper for a `FILE *` that handles opening and closing a file in RAII
+/// fashion. Errors are handled by raising instances of
+/// `std::filesystem::filesystem_error` so that there's a single exception type
+/// being raised for all filesystem-related errors.
+class File {
+public:
+    /// Creates an instance of `File` without opening any file.
+    File() = default;
+
+    /// Creates an instance of `File` and opens the file at `path` using mode `mode`.
+    ///
+    /// @param path The path of the file.
+    /// @param mode The mode for opening the file, passed to `fdopen()`.
+    File(const std::filesystem::path & path, const char * mode);
+
+    /// Creates an instance of `File` by opening file descriptor `fd`. The
+    /// `path` argument is only used for error reporting and for retrieving from
+    /// the object via `get_path()`.
+    ///
+    /// @param fd The file descriptor to call fdopen() on.
+    /// @param path The path of the file, for error reporting etc.
+    /// @param mode The mode for opening the file, passed to `fdopen()`.
+    File(int fd, const std::filesystem::path & path, const char * mode);
+
+    File(const File &) = delete;
+    File & operator=(const File &) = delete;
+    File(File &&) = default;
```

Comment:
It is not a good idea to use the default implementation of move operations. `file` will not be set to nullptr in the source. There will be a double free error.